src/ngraph/runtime/reference/topk.hpp

```diff
@@ -0,0 +1,95 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <cmath>
+#include <algorithm>
+#include <numeric>
+
+#include "ngraph/coordinate_transform.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            template <typename T, typename U>
+            void topk(
+                const T* arg, U* out_indices, T* out_values, const Shape& in_shape, const Shape& out_shape, size_t axis, size_t k, bool compute_max)
+            {
+                // reorder source axis visit order and make "axis" inner most
+                size_t ndim = static_cast<size_t>(in_shape.size());
+                Coordinate start_corner(ndim, 0);
+                Coordinate end_corner(in_shape);
+                end_corner[axis] = 1;
+                Strides strides(ndim, 1);
+                AxisVector axis_order(ndim);
+                std::iota(axis_order.begin(), axis_order.end(), 0);
+                axis_order.erase(axis_order.begin() + axis);
+                axis_order.push_back(axis);
+                // Create CoordinateTransforms that visits only the first element along "axis"
+                CoordinateTransform input_transform(in_shape,
+                        start_corner,
+                        end_corner,
+                        strides,
+                        axis_order);
+                CoordinateTransform output_transform(out_shape,
+                        start_corner,
+                        end_corner,
+                        strides,
+                        axis_order);
+                // Create temp vector for sorting.
+                std::vector<std::tuple<T, U>> workspace(in_shape[axis]);
+                std::vector<size_t> in_strides = ngraph::row_major_strides(in_shape);
+                std::vector<size_t> out_strides = ngraph::row_major_strides(out_shape);
+                auto in_axis_stride = in_strides[axis];
+                auto out_axis_stride = out_strides[axis];
+                for(const Coordinate& coord: input_transform)
+                {
+                    auto arg_index = input_transform.index(coord);
+                    auto out_index = output_transform.index(coord);
+                    // Fill the temp vector
+                    U i = 0;
+                    for(std::tuple<T, U> &entry : workspace)
+                    {
+                        // std::cout << "Read arg[" << arg_index << "]" <<std::endl;
+                        std::get<0>(entry) = arg[arg_index];
+                        std::get<1>(entry) = i;
+                        arg_index += in_axis_stride;
+                        i++;
+                    }
+                    // Sort the temp vector
+                    std::sort(workspace.begin(),
+                            workspace.end(),
+                            compute_max ?
```

Comment:
If we compare with the ordering on the tuple, then in case of ties we're going to wind up with a situation where for `compute_max` the highest-indexed element is going to win, while for `compute_min` the lowest-indexed element is going to win. I think for the reference the best thing is to have the lowest-indexed element always win, so something like:

```
[] (const std::tuple<T,U>& a, const std::tuple<T,U>& b) -> bool { return a.first == b.first ? a.second < b.second : a.first < b.first; } :
[] (const std::tuple<T,U>& a, const std::tuple<T,U>& b) -> bool { return a.first == b.first ? a.second < b.second : a.first > b.first; }
```